src/TermInterface.jl

```diff
@@ -32,43 +46,40 @@ issym(x) = false
 export issym
 
 """
-  exprhead(x)
-
-If `x` is a term as defined by `istree(x)`, `exprhead(x)` must return a symbol,
-corresponding to the head of the `Expr` most similar to the term `x`.
-If `x` represents a function call, for example, the `exprhead` is `:call`.
-If `x` represents an indexing operation, such as `arr[i]`, then `exprhead` is `:ref`.
-Note that `exprhead` is different from `operation` and both functions should 
-be defined correctly in order to let other packages provide code generation 
-and pattern matching features. 
+    head(x)
+Returns the head of the S-expression.
 """
-function exprhead end
-export exprhead
+function head end
 
+"""
+    children(x)
+Returns the children (aka tail) of the S-expression.
+"""
+function children end
 
 """
   operation(x)
 
-If `x` is a term as defined by `istree(x)`, `operation(x)` returns the
-head of the term if `x` represents a function call, for example, the head
-is the function being called.
+Returns the function a function call expression is calling.
+`iscall(x)` must be true as a precondition.
 """
-function operation end
+operation(x) = iscall(x) ? first(children(x)) : error("operation called on a non-function call expression")
```

Comment:
Ha! @shashi  I would leave this undefined, `function operation end`. This default expr-like behavior of operation being `first(children(x))` is only relevant to `Expr`, there may be some `AST` that just has `operation(x) = head(x)` (e.g. Symbolics.jl Term)

Reply:
Agreed. Let's define operation and arguments explicitly for each term type.